viewing_party/party.py

```diff
@@ -1,23 +1,112 @@
+def get_watched_avg_rating(user_data):
+    ratings = []
+    average_rating = 0
+    for movie in user_data['watched']:
+        ratings.append(movie['rating'])
+    if len(ratings) > 0:
+        average_rating = sum(ratings)/len(ratings)
+    else:
+        average_rating = sum(ratings)
+    return average_rating
 
+def get_most_watched_genre(user_data):
+    genres = []
+    # add all user's genres to a list
+    for movie in user_data["watched"]:
+        if 'genre' in movie:
+            genres.append(movie['genre'])
+    # find user's most watched genre
+    if len(genres) == 0:
+        return None
+    else:
+        most_watched_genre = None
+        counter = 0
+        for genre in genres:
+            curr_frequency = genres.count(genre)
+            if (curr_frequency > counter):
+                counter = curr_frequency
+                most_watched_genre = genre
+    return most_watched_genre
 
+    return popular_genre
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    friend_movies = []
+    unique_user_movies = []
+    # generate friend watched movies list
+    for friend_data in user_data["friends"]:
+        if 'watched' in friend_data:
+            for movie in friend_data['watched']:
+                if movie not in friend_movies:
+                    friend_movies.append(movie)
+    # find users unique movies
+    for movie in user_data["watched"]:
+        if movie not in friend_movies and movie not in unique_user_movies:
+            unique_user_movies.append(movie)
+    return unique_user_movies
+
+def get_friends_unique_watched(user_data):
+    user_movies = user_data["watched"]
+    unique_friend_movies = []
+    for dict in user_data["friends"]:
+        if 'watched' in dict:
+            for movie in dict['watched']:
+                if movie not in user_movies and movie not in unique_friend_movies:
+                    unique_friend_movies.append(movie)
+    return unique_friend_movies
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    unique_friend_movies = get_friends_unique_watched(user_data)
+    user_subscriptions = user_data['subscriptions']
+    recommendations = []
+    for movie in unique_friend_movies:
+        if 'host' in movie and movie['host'] in user_subscriptions:
+            recommendations.append(movie)
+    return recommendations
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    unique_friend_movies = get_friends_unique_watched(user_data)
+    most_watched_genre = get_most_watched_genre(user_data)
+    return [movie for movie in unique_friend_movies if movie['genre'] == most_watched_genre]

```

Comment:
Very elegant list comprehension and function as a whole, supported by helper functions!